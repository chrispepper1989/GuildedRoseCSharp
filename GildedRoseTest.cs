using System;
using System.IO;
using System.Text;
using ApprovalTests;
using ApprovalTests.Reporters;
using csharp;

namespace GildedRoseCSharp
{
  
    [UseReporter(typeof(DiffReporter))]
    [TestFixture]
    public class GildedRoseTest
    {

        //generate a bunch of test cases
        public IList<Item> GenerateTestItems(ValidItems[] names, int[] sellIns, int[] qualities)
        {
            IList<Item> Items = new List<Item>
            {
            };
            foreach (var name in names)
            {
                foreach (var sellIn in sellIns)
                {
                    foreach (var quality in qualities)
                    {
                        var item = ItemHelper.CreateItem(name, sellIn, quality);
                        Items.Add( item);
                    }

                }

            }

            return Items;

        }

        //[TestCase(10)] //todo why doesnt this work with Approvals?
        public void TestUpdateQuality(int days)
        {

            //arrange
            ValidItems[] namesProper = new[]
            {
                ValidItems.AgedBrie, ValidItems.DexterityVest, ValidItems.ElixirOfMongoose,
                ValidItems.SulfurasHandOfRagnaros, ValidItems.ConjuredManaCake
            };
            var sellIns = new[] { 1, 5, 10, 2, 4 };
            var qualities = new[] { 0, 1, 2, 5, 29 };
            var Items = GenerateTestItems(namesProper, sellIns, qualities);
            GildedRose app = new GildedRose(Items);
            //act
            for (var i = 0; i < days; ++i)
            {
                app.UpdateQuality();
            }
            //assert
            Approvals.VerifyAll(Items, "All Items");

        }
        [Test]
        public void ConjuredManaCakeDegradesTwiceAsFastAsNormalItems()
        {
            //arrange
            var conitem = ItemHelper.CreateItem(ValidItems.ConjuredManaCake, 10, 100);
            var normalItem = ItemHelper.CreateItem(ValidItems.ElixirOfMongoose, 10, 100);

            IList<Item> Items = new List<Item>
            {
         
            };

            GildedRose app = new GildedRose(Items);

            IList<Item> ItemsOverTime = new List<Item>();
            //act
            for (var i = 0; i < 100; ++i)
            {
                conitem.SellIn -= 1;
                normalItem.SellIn -= 1;
                //normal item done twice
                normalItem.Quality =  GildedRose.GetNewStandardItemQuality(normalItem);
                normalItem.Quality = GildedRose.GetNewStandardItemQuality(normalItem);
                conitem.Quality = GildedRose.GetNewConjuredManaCakeQuality(conitem);
               
                //assert
                Assert.AreEqual(normalItem.Quality, conitem.Quality);
      
                ItemsOverTime.Add(new Item()
                {
                    Name = normalItem.Name,
                    Quality = normalItem.Quality,
                    SellIn = normalItem.SellIn

                });
                ItemsOverTime.Add(new Item()
                {
                    Name = conitem.Name,
                    Quality = conitem.Quality,
                    SellIn = conitem.SellIn

                });

            }
            Approvals.VerifyAll(ItemsOverTime, "Conjured Items ");

        }

        [Test]
         public void TestUpdateQuality_1Days()
         {
             TestUpdateQuality(1);
         }
         [Test]
         public void TestUpdateQuality_5Days()
         {
             TestUpdateQuality(5);
         }
         [Test]
         public void TestUpdateQuality_10Days()
         {
             TestUpdateQuality(10);
         }
         [Test]
         public void TestUpdateQuality_30Days()
         {
             TestUpdateQuality(30);
         }

    }
}
