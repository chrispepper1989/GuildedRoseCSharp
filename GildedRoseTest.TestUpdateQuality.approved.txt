All Items[0] = Aged Brie, 0, 1
All Items[1] = Aged Brie, 0, 2
All Items[2] = Aged Brie, 0, 3
All Items[3] = Aged Brie, 0, 6
All Items[4] = Aged Brie, 0, 8
All Items[5] = Aged Brie, 0, 21
All Items[6] = Aged Brie, 0, 80
All Items[7] = Aged Brie, 4, 1
All Items[8] = Aged Brie, 4, 2
All Items[9] = Aged Brie, 4, 3
All Items[10] = Aged Brie, 4, 6
All Items[11] = Aged Brie, 4, 8
All Items[12] = Aged Brie, 4, 21
All Items[13] = Aged Brie, 4, 80
All Items[14] = Aged Brie, 9, 1
All Items[15] = Aged Brie, 9, 2
All Items[16] = Aged Brie, 9, 3
All Items[17] = Aged Brie, 9, 6
All Items[18] = Aged Brie, 9, 8
All Items[19] = Aged Brie, 9, 21
All Items[20] = Aged Brie, 9, 80
All Items[21] = Aged Brie, 1, 1
All Items[22] = Aged Brie, 1, 2
All Items[23] = Aged Brie, 1, 3
All Items[24] = Aged Brie, 1, 6
All Items[25] = Aged Brie, 1, 8
All Items[26] = Aged Brie, 1, 21
All Items[27] = Aged Brie, 1, 80
All Items[28] = Aged Brie, 3, 1
All Items[29] = Aged Brie, 3, 2
All Items[30] = Aged Brie, 3, 3
All Items[31] = Aged Brie, 3, 6
All Items[32] = Aged Brie, 3, 8
All Items[33] = Aged Brie, 3, 21
All Items[34] = Aged Brie, 3, 80
All Items[35] = +5 Dexterity Vest, 0, 0
All Items[36] = +5 Dexterity Vest, 0, 0
All Items[37] = +5 Dexterity Vest, 0, 1
All Items[38] = +5 Dexterity Vest, 0, 4
All Items[39] = +5 Dexterity Vest, 0, 6
All Items[40] = +5 Dexterity Vest, 0, 19
All Items[41] = +5 Dexterity Vest, 0, 79
All Items[42] = +5 Dexterity Vest, 4, 0
All Items[43] = +5 Dexterity Vest, 4, 0
All Items[44] = +5 Dexterity Vest, 4, 1
All Items[45] = +5 Dexterity Vest, 4, 4
All Items[46] = +5 Dexterity Vest, 4, 6
All Items[47] = +5 Dexterity Vest, 4, 19
All Items[48] = +5 Dexterity Vest, 4, 79
All Items[49] = +5 Dexterity Vest, 9, 0
All Items[50] = +5 Dexterity Vest, 9, 0
All Items[51] = +5 Dexterity Vest, 9, 1
All Items[52] = +5 Dexterity Vest, 9, 4
All Items[53] = +5 Dexterity Vest, 9, 6
All Items[54] = +5 Dexterity Vest, 9, 19
All Items[55] = +5 Dexterity Vest, 9, 79
All Items[56] = +5 Dexterity Vest, 1, 0
All Items[57] = +5 Dexterity Vest, 1, 0
All Items[58] = +5 Dexterity Vest, 1, 1
All Items[59] = +5 Dexterity Vest, 1, 4
All Items[60] = +5 Dexterity Vest, 1, 6
All Items[61] = +5 Dexterity Vest, 1, 19
All Items[62] = +5 Dexterity Vest, 1, 79
All Items[63] = +5 Dexterity Vest, 3, 0
All Items[64] = +5 Dexterity Vest, 3, 0
All Items[65] = +5 Dexterity Vest, 3, 1
All Items[66] = +5 Dexterity Vest, 3, 4
All Items[67] = +5 Dexterity Vest, 3, 6
All Items[68] = +5 Dexterity Vest, 3, 19
All Items[69] = +5 Dexterity Vest, 3, 79
All Items[70] = Elixir of the Mongoose, 0, 0
All Items[71] = Elixir of the Mongoose, 0, 0
All Items[72] = Elixir of the Mongoose, 0, 1
All Items[73] = Elixir of the Mongoose, 0, 4
All Items[74] = Elixir of the Mongoose, 0, 6
All Items[75] = Elixir of the Mongoose, 0, 19
All Items[76] = Elixir of the Mongoose, 0, 79
All Items[77] = Elixir of the Mongoose, 4, 0
All Items[78] = Elixir of the Mongoose, 4, 0
All Items[79] = Elixir of the Mongoose, 4, 1
All Items[80] = Elixir of the Mongoose, 4, 4
All Items[81] = Elixir of the Mongoose, 4, 6
All Items[82] = Elixir of the Mongoose, 4, 19
All Items[83] = Elixir of the Mongoose, 4, 79
All Items[84] = Elixir of the Mongoose, 9, 0
All Items[85] = Elixir of the Mongoose, 9, 0
All Items[86] = Elixir of the Mongoose, 9, 1
All Items[87] = Elixir of the Mongoose, 9, 4
All Items[88] = Elixir of the Mongoose, 9, 6
All Items[89] = Elixir of the Mongoose, 9, 19
All Items[90] = Elixir of the Mongoose, 9, 79
All Items[91] = Elixir of the Mongoose, 1, 0
All Items[92] = Elixir of the Mongoose, 1, 0
All Items[93] = Elixir of the Mongoose, 1, 1
All Items[94] = Elixir of the Mongoose, 1, 4
All Items[95] = Elixir of the Mongoose, 1, 6
All Items[96] = Elixir of the Mongoose, 1, 19
All Items[97] = Elixir of the Mongoose, 1, 79
All Items[98] = Elixir of the Mongoose, 3, 0
All Items[99] = Elixir of the Mongoose, 3, 0
All Items[100] = Elixir of the Mongoose, 3, 1
All Items[101] = Elixir of the Mongoose, 3, 4
All Items[102] = Elixir of the Mongoose, 3, 6
All Items[103] = Elixir of the Mongoose, 3, 19
All Items[104] = Elixir of the Mongoose, 3, 79
All Items[105] = Sulfuras, Hand of Ragnaros, 1, 0
All Items[106] = Sulfuras, Hand of Ragnaros, 1, 1
All Items[107] = Sulfuras, Hand of Ragnaros, 1, 2
All Items[108] = Sulfuras, Hand of Ragnaros, 1, 5
All Items[109] = Sulfuras, Hand of Ragnaros, 1, 7
All Items[110] = Sulfuras, Hand of Ragnaros, 1, 20
All Items[111] = Sulfuras, Hand of Ragnaros, 1, 80
All Items[112] = Sulfuras, Hand of Ragnaros, 5, 0
All Items[113] = Sulfuras, Hand of Ragnaros, 5, 1
All Items[114] = Sulfuras, Hand of Ragnaros, 5, 2
All Items[115] = Sulfuras, Hand of Ragnaros, 5, 5
All Items[116] = Sulfuras, Hand of Ragnaros, 5, 7
All Items[117] = Sulfuras, Hand of Ragnaros, 5, 20
All Items[118] = Sulfuras, Hand of Ragnaros, 5, 80
All Items[119] = Sulfuras, Hand of Ragnaros, 10, 0
All Items[120] = Sulfuras, Hand of Ragnaros, 10, 1
All Items[121] = Sulfuras, Hand of Ragnaros, 10, 2
All Items[122] = Sulfuras, Hand of Ragnaros, 10, 5
All Items[123] = Sulfuras, Hand of Ragnaros, 10, 7
All Items[124] = Sulfuras, Hand of Ragnaros, 10, 20
All Items[125] = Sulfuras, Hand of Ragnaros, 10, 80
All Items[126] = Sulfuras, Hand of Ragnaros, 2, 0
All Items[127] = Sulfuras, Hand of Ragnaros, 2, 1
All Items[128] = Sulfuras, Hand of Ragnaros, 2, 2
All Items[129] = Sulfuras, Hand of Ragnaros, 2, 5
All Items[130] = Sulfuras, Hand of Ragnaros, 2, 7
All Items[131] = Sulfuras, Hand of Ragnaros, 2, 20
All Items[132] = Sulfuras, Hand of Ragnaros, 2, 80
All Items[133] = Sulfuras, Hand of Ragnaros, 4, 0
All Items[134] = Sulfuras, Hand of Ragnaros, 4, 1
All Items[135] = Sulfuras, Hand of Ragnaros, 4, 2
All Items[136] = Sulfuras, Hand of Ragnaros, 4, 5
All Items[137] = Sulfuras, Hand of Ragnaros, 4, 7
All Items[138] = Sulfuras, Hand of Ragnaros, 4, 20
All Items[139] = Sulfuras, Hand of Ragnaros, 4, 80
All Items[140] = Sulfuras, Hand of Ragnaros, 1, 0
All Items[141] = Sulfuras, Hand of Ragnaros, 1, 1
All Items[142] = Sulfuras, Hand of Ragnaros, 1, 2
All Items[143] = Sulfuras, Hand of Ragnaros, 1, 5
All Items[144] = Sulfuras, Hand of Ragnaros, 1, 7
All Items[145] = Sulfuras, Hand of Ragnaros, 1, 20
All Items[146] = Sulfuras, Hand of Ragnaros, 1, 80
All Items[147] = Sulfuras, Hand of Ragnaros, 5, 0
All Items[148] = Sulfuras, Hand of Ragnaros, 5, 1
All Items[149] = Sulfuras, Hand of Ragnaros, 5, 2
All Items[150] = Sulfuras, Hand of Ragnaros, 5, 5
All Items[151] = Sulfuras, Hand of Ragnaros, 5, 7
All Items[152] = Sulfuras, Hand of Ragnaros, 5, 20
All Items[153] = Sulfuras, Hand of Ragnaros, 5, 80
All Items[154] = Sulfuras, Hand of Ragnaros, 10, 0
All Items[155] = Sulfuras, Hand of Ragnaros, 10, 1
All Items[156] = Sulfuras, Hand of Ragnaros, 10, 2
All Items[157] = Sulfuras, Hand of Ragnaros, 10, 5
All Items[158] = Sulfuras, Hand of Ragnaros, 10, 7
All Items[159] = Sulfuras, Hand of Ragnaros, 10, 20
All Items[160] = Sulfuras, Hand of Ragnaros, 10, 80
All Items[161] = Sulfuras, Hand of Ragnaros, 2, 0
All Items[162] = Sulfuras, Hand of Ragnaros, 2, 1
All Items[163] = Sulfuras, Hand of Ragnaros, 2, 2
All Items[164] = Sulfuras, Hand of Ragnaros, 2, 5
All Items[165] = Sulfuras, Hand of Ragnaros, 2, 7
All Items[166] = Sulfuras, Hand of Ragnaros, 2, 20
All Items[167] = Sulfuras, Hand of Ragnaros, 2, 80
All Items[168] = Sulfuras, Hand of Ragnaros, 4, 0
All Items[169] = Sulfuras, Hand of Ragnaros, 4, 1
All Items[170] = Sulfuras, Hand of Ragnaros, 4, 2
All Items[171] = Sulfuras, Hand of Ragnaros, 4, 5
All Items[172] = Sulfuras, Hand of Ragnaros, 4, 7
All Items[173] = Sulfuras, Hand of Ragnaros, 4, 20
All Items[174] = Sulfuras, Hand of Ragnaros, 4, 80
